libs/agno/agno/tools/function.py

```diff
@@ -548,6 +594,77 @@ async def _handle_post_hook_async(self):
                 log_warning(f"Error in post-hook callback: {e}")
                 log_exception(e)
 
+    async def _build_nested_execution_chain_async(self, entrypoint_args: Dict[str, Any]):
+        """Build a nested chain of async hook executions with the entrypoint at the center.
+
+        Similar to _build_nested_execution_chain but for async execution.
+        """
+        from functools import reduce
+        from inspect import isasyncgen, isasyncgenfunction, iscoroutinefunction
+
+        # Start with the innermost function execution
+        async def execute_entrypoint_async(name, func, args):
+            arguments = entrypoint_args.copy()
+            if self.arguments is not None:
+                arguments.update(self.arguments)
+
+            result = self.function.entrypoint(**arguments)  # type: ignore
+            if iscoroutinefunction(self.function.entrypoint) and not (
+                isasyncgen(self.function.entrypoint) or isasyncgenfunction(self.function.entrypoint)
+            ):
+                result = await result
+            return result
+
+        def execute_entrypoint(name, func, args):
+            arguments = entrypoint_args.copy()
+            if self.arguments is not None:
+                arguments.update(self.arguments)
+            return self.function.entrypoint(**arguments)  # type: ignore
```

Comment:
if we define this a bit higher we can use it in the execute() method too and remove the duplication